src/peak_finding/1d_list.clj

```diff
@@ -0,0 +1,82 @@
+;; A peak is any element that is larger or equal to both it's neighbours
+;; Items that only have one neighbour (end items) are considered a peak if
+;; they are larger or equal to their neighbour
+;; This algorithm returns the first peak it finds.
+
+(ns peakfinder)
+
+(defn ^:dynamic find-peak
+  "multiple arity recursive function, for finding a peak in a list"
+  ([list] (->> (middle-val list) (item list) (find-peak list)))
+
+  ([list current-item]
+   (if (empty? list) -1
+       (let [index (:index current-item)
+             left (left-neighbour list current-item)
+             right (right-neighbour list current-item)]
+         (cond
+           (-> list (count) (- 1) (= index)) ;; Furthest item right
+           (peak? current-item left)
+
+           (= index 0) ;; Furthest item left
+           (peak? current-item right)
+
+           :else
+           (if (= (peak? current-item left right) -1)
+
+             (if (> (:value left) (:value current-item))
+               (find-peak list left) ;; Recursive
+               (find-peak list right)) ;; Recursive
+
+             (:value current-item)))))))
+
+(defn- ^:dynamic peak?
+  "determines if an item is a peak compared to the supplied comparitors"
+  [item & comparitors]
+  (let [item-value (:value item)
+        values (map (fn [item] (:value item)) comparitors)]
+    (if (every? (fn [v] (>= item-value v)) values) item-value -1)))
+
+(defn- item
+  "returns a map that represents an item in a list. The data describes the
+  index at which the item resides in the list and it's value. e.g. {index 1 :value 3}"
+  [list, index]
+  {:index index :value (nth list index nil)})
```

Comment:
This is potentially dangerous performance wise if list is a list or a seq or something that does not have random access.  See comment about this below.  One general approach to solving this would be to convert your input list to a vector at the very beginning of the algorithm and ensure that you use vectors throughout.  Vectors support fast access to particular indexes.